tls/s2n_resume.c

```diff
@@ -585,6 +585,116 @@ int s2n_decrypt_session_ticket(struct s2n_connection *conn)
     return 0;
 }
 
+/* session is the original session data, out is the encrypted output bytes, including 
+ * the key_name, iv, and the encrypted session data
+ */
+int s2n_encrypt_session_cache(struct s2n_connection *conn, const uint8_t *session, uint8_t *out)
+{
+    struct s2n_ticket_key *key;
+    struct s2n_session_key aes_ticket_key;
+    struct s2n_blob aes_key_blob;
+
+    uint8_t iv_data[S2N_TLS_GCM_IV_LEN] = { 0 };
+    struct s2n_blob iv = { .data = iv_data, .size = sizeof(iv_data) };
+
+    uint8_t aad_data[S2N_TICKET_AAD_LEN] = { 0 };
+    struct s2n_blob aad_blob = { .data = aad_data, .size = sizeof(aad_data) };
+    struct s2n_stuffer aad;
+
+    uint8_t s_data[S2N_STATE_SIZE_IN_BYTES + S2N_TLS_GCM_TAG_LEN] = { 0 };
+    struct s2n_blob state_blob = { .data = s_data, .size = S2N_STATE_SIZE_IN_BYTES + S2N_TLS_GCM_TAG_LEN };
+    struct s2n_stuffer state;
+
+    GUARD(s2n_stuffer_init(&state, &state_blob));
+    GUARD(s2n_stuffer_write_bytes(&state, session, S2N_STATE_SIZE_IN_BYTES));
+
+    key = s2n_get_ticket_encrypt_decrypt_key(conn->config);
+
+    /* No keys loaded by the user or the keys are either in decrypt-only or expired state */
+    S2N_ERROR_IF(!key, S2N_ERR_NO_TICKET_ENCRYPT_DECRYPT_KEY);
+
+    GUARD(s2n_get_public_random_data(&iv));
+
+    s2n_blob_init(&aes_key_blob, key->aes_key, S2N_AES256_KEY_LEN);
+    notnull_check(aes_ticket_key.evp_cipher_ctx = EVP_CIPHER_CTX_new());
+    GUARD(s2n_aes256_gcm.init(&aes_ticket_key));
+    GUARD(s2n_aes256_gcm.set_encryption_key(&aes_ticket_key, &aes_key_blob));
+
+    GUARD(s2n_stuffer_init(&aad, &aad_blob));
+    GUARD(s2n_stuffer_write_bytes(&aad, key->implicit_aad, S2N_TICKET_AAD_IMPLICIT_LEN));
+    GUARD(s2n_stuffer_write_bytes(&aad, key->key_name, S2N_TICKET_KEY_NAME_LEN));
+
+    GUARD(s2n_aes256_gcm.io.aead.encrypt(&aes_ticket_key, &iv, &aad_blob, &state_blob, &state_blob));
+
+    memcpy_check(out, key->key_name, S2N_TICKET_KEY_NAME_LEN);
+    out += S2N_TICKET_KEY_NAME_LEN;
+    memcpy_check(out, iv_data, S2N_TLS_GCM_IV_LEN);
+    out += S2N_TLS_GCM_IV_LEN;
+    memcpy_check(out, s_data, S2N_STATE_SIZE_IN_BYTES + S2N_TLS_GCM_TAG_LEN);
+
+    GUARD(s2n_aes256_gcm.destroy_key(&aes_ticket_key));
+    GUARD(s2n_session_key_free(&aes_ticket_key));
+
+    return 0;
+}
```

Comment:
How do we change the encrypted blob format? Should there be a version?

Are there any tests for these new functions?

Reply:
yeah the UTs are here: https://github.com/awslabs/s2n/pull/902/commits/c9925c394314c00951dc51adaa1fa2babd8f64ba#diff-ab06370f5023bc039f123d47bb7353f7

What do you mean version and "change the encrypted blob format"? Can you explain more about them?

Reply:
discussed with our team, I think this is more than just adding a text version number. Since right now session ticket also doesn't have a version number so it makes more sense to add this version for session cache and session ticket together. Or more generically, whenever we encrypt some data blob we need this. 
So we can consider this in a different PR. And our team can work on it later. 
This PR is only to use the same session ticket key to encrypt the session cache data.

Reply:
By reading the encryption/decryption code, I do believe session ticket key rotation is already considered as KEY_NAME is embeded into encoded session data. The API is designed in a way for caller to use. I am changing it to be called inside cache_stored/cache_retreive, as encryption/decryption is more like a native requirement for s2n to handle session data. A new PR would be pushed